onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -4,213 +4,577 @@
-                                                        const TensorShape& tensor_shape,
-                                                        std::unique_ptr<Tensor>* p_tensor,
-                                                        AllocatorPtr alloc,
-                                                        void* preallocated,
-                                                        size_t preallocated_size) {
+#define CASE_PROTO_TRACE(X, Y)                                                            \
+  case ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_##X:                    \
+    if (!IAllocator::CalcMemSizeForArrayWithAlignment<alignment>(size, sizeof(Y), out)) { \
+      return common::Status(common::ONNXRUNTIME, common::FAIL, "Invalid TensorProto");    \
+    }                                                                                     \
+    break;
+
+template <size_t alignment>
+common::Status GetSizeInBytesFromTensorProto(const ONNX_NAMESPACE::TensorProto& tensor_proto, size_t* out) {
+  const auto& dims = tensor_proto.dims();
+  size_t size = 1;
+  for (int i = 0; i < dims.size(); ++i) {
+    if (dims[i] < 0) {
+      return common::Status(common::ONNXRUNTIME, common::FAIL, "Invalid TensorProto");
+    }
+    if (!IAllocator::CalcMemSizeForArray(size, static_cast<size_t>(dims[i]), &size)) {
+      return common::Status(common::ONNXRUNTIME, common::FAIL, "Invalid TensorProto");
+    }
+  }
+  switch (tensor_proto.data_type()) {
+    CASE_PROTO_TRACE(FLOAT, float);
+    CASE_PROTO_TRACE(DOUBLE, double);
+    CASE_PROTO_TRACE(BOOL, bool);
+    CASE_PROTO_TRACE(INT8, int8_t);
+    CASE_PROTO_TRACE(INT16, int16_t);
+    CASE_PROTO_TRACE(INT32, int32_t);
+    CASE_PROTO_TRACE(INT64, int64_t);
+    CASE_PROTO_TRACE(UINT8, uint8_t);
+    CASE_PROTO_TRACE(UINT16, uint16_t);
+    CASE_PROTO_TRACE(UINT32, uint32_t);
+    CASE_PROTO_TRACE(UINT64, uint64_t);
+    CASE_PROTO_TRACE(FLOAT16, MLFloat16);
+    CASE_PROTO_TRACE(BFLOAT16, BFloat16);
+    CASE_PROTO_TRACE(STRING, std::string);
+    default:
+      return common::Status(common::ONNXRUNTIME, common::NOT_IMPLEMENTED);
+  }
+  return Status::OK();
+}
+
+std::vector<int64_t> GetTensorShapeFromTensorShapeProto(const ONNX_NAMESPACE::TensorShapeProto& tensor_shape_proto) {
+  const auto& dims = tensor_shape_proto.dim();
+  std::vector<int64_t> tensor_shape_vec(static_cast<size_t>(dims.size()));
+  for (int i = 0; i < dims.size(); ++i) {
+    tensor_shape_vec[i] = dims[i].has_dim_param() ? -1 /* symbolic dimensions are represented as -1 in onnxruntime*/
+                                                  : dims[i].dim_value();
+  }
+  return tensor_shape_vec;
+}
+
+template <typename T>
+common::Status GetTensorByTypeFromTensorProto(const TensorProto& tensor_proto, const TensorShape& tensor_shape,
+                                              std::unique_ptr<Tensor>* p_tensor, const OrtAllocatorInfo& alloc,
```

Comment:
>>* p_tensor [](start = 68, length = 11)

I know this was existing code, but unique_ptr<Tensor>& would be more consistent with the latest code for a parameter that can't be null.